### DISCUSSION TITLE
why can't c++ Open3D Application “find resource directory”?

### DISCUSSION THREAD

**Discussion by cwreynolds** [isl-org/Open3D#7000] [Q&A] [answered]
This is a repost of a question I asked a few days ago on stackoverflow (see [here](https://stackoverflow.com/questions/79048820/why-cant-c-open3d-application-find-resource-directory) and below). Any hints or suggestions would be greatly appreciated. My main question below is “What is the resource directory, what needs to be in it, and how do I set that up?”

I came across [Open3DViewer.cpp](https://github.com/isl-org/Open3D/blob/main/cpp/apps/Open3DViewer/Open3DViewer.cpp) which calls `open3d::visualization::app::RunViewer()`. This might be the sort of example I was looking for. But aside from the bare minimum [description in the API doc](https://www.open3d.org/docs/latest/cpp_api/namespaceopen3d_1_1visualization_1_1app.html), there is no discussion of “resource” requirements, and no instructions for a building (say) a copy of the Open3DViewer directory as a starting point for a new app project.

> .
> I am using Open3D (0.18.0+) from C++, and having some troubles adjusting properties in `ViewControl`.
> 
> I have been using `open3d::visualization::Visualizer` — which is apparently considered “legacy” — versus say `open3d::visualization::gui::Application`. I would prefer to use the newer `Application` class, but I have problems with that too.
> 
> This code using the legacy `Visualizer` works, showing a window with a red ball in the center:
> 
> ```c++
> void test()
> {
>     open3d::PrintOpen3DVersion();
>     auto vis = open3d::visualization::Visualizer();
>     vis.CreateVisualizerWindow();
>     auto ball = open3d::geometry::TriangleMesh::CreateSphere(5);
>     ball->ComputeVertexNormals();
>     ball->PaintUniformColor({1, 0, 0});
>     vis.AddGeometry(ball);
>     vis.Run();
> }
> ```
> 
> While there seems to be very little Open3D example code in C++, there is a fair amount in Python. Most of it uses the modern `Application` class. Not wanting to go against the flow, I tried something similar using that newer class:
> 
> ```c++
> void test2()
> {
>     
>     open3d::PrintOpen3DVersion();
>     open3d::visualization::gui::Application::GetInstance().Initialize();
>     auto vis = open3d::visualization::visualizer::O3DVisualizer("name", 500, 500);
>     auto ball = open3d::geometry::TriangleMesh::CreateSphere(5);
>     ball->ComputeVertexNormals();
>     ball->PaintUniformColor({1, 0, 0});
>     vis.AddGeometry("ball", {ball});
>     open3d::visualization::gui::Application::GetInstance().Run();
> }
> ```
> 
> That gets “Could not find resource directory”:
> 
> ```
> libc++abi: terminating due to uncaught exception of type std::runtime_error:
> [Open3D Error] (std::string (anonymous namespace)::FindResourcePath(int, 
> const char **)) 
> /Users/cwr/Documents/code/Open3D/cpp/open3d/visualization/gui/
> Application.cpp:96: Could not find resource directory.
> ```
> 
> What is the resource directory, what needs to be in it, and how do I set that up?
> 
> It would be super helpful to see modern example C++ code using the `Application` class.
>

**Comment by cwreynolds** [comment]
Looking closer, I notice that [Open3DViewer.cpp](https://github.com/isl-org/Open3D/blob/main/cpp/apps/Open3DViewer/Open3DViewer.cpp) calls `RunViewer()`, except when compiled for `__APPLE__`, in which case it does nothing. That branch of the `#if` has a commented out mention of `Open3DViewer_mac.mm` an Objective-C file, perhaps now obsolete? 

So it looks like that cannot serve as a model for the work I am doing on macOS. I continue to seek information on running a program like `test2()` above using `open3d::visualization::gui::Application`.

**Comment by ssheorey** [comment, marked as answer]
`examples/cpp/Draw.cpp` uses a simplified interface over O3DVIsualizer and may be sufficient for your purpose.
Open3DViewer_mac.mm is the right example for Apple. Thanks for catching that its commented out in the C++ code. The build system directly builds that now, instead of going through Open3DVIewer.cpp for macOS. (`cpp/apps/CMakeLists.txt`)

The "resources not found"  error is since we use some heuristics to locate the resource directory. See the function `FindResourcePath()` in `cpp/open3d/visualization/gui/Application.cpp`. 

Ideally this would be relative to the executable location. [Looking for cross-platform PRs here :-) ]

**Reply by cwreynolds** [reply]
Thanks for all that! I marked it as answered. As I dig in there will probably be more noob questions later.

**Reply by cwreynolds** [reply]
I looked at `examples/cpp/Draw.cpp` and yes, that seems a simple enough place to start. And in fact I had found it in my earlier wandering through the Open3D sources. Back in August (#6923) I was trying to link my code to Open3D. I copied the contents of `Draw.cpp` into my main.cpp just so I could build/link/run something. I figured I was doing it wrong because I got “Could not find resource directory.”

Now following suggestions from @ssheorey I tried to decide how to build `Draw.cpp`. Looking through the Open3D build directory I found it was already built: `build/bin/examples/Draw`. But dragging that executable to a Terminal window, or double clicking on it, gets the same “Could not find resource directory” error.

After further poking around inside `build/bin` I found the main `Open3D.app` next to a directory called `Resources`. Is that in fact the elusive “resource directory”, that I keep being unable to find?! 

I see that `Draw.cpp` does not call `FindResourcePath()` so I don't know how to tell it where to find the resource directory. Might it work to place my executable (or say `build/bin/examples/Draw`) in a directory along with a _copy_ of `build/bin/Resources`?

[I look forward to eventually knowing enough to be able to contribute “cross-platform PRs” here, but first I have to understand it well enough to get it to run.]

**Reply by cwreynolds** [reply]
> ...Might it work to place my executable (or say `build/bin/examples/Draw`) in a directory along with a _copy_ of `build/bin/Resources`?...

Sadly, the answer is **no**. I had hoped this might provide a simple work-around. I made a temp dir, dragged in copies of `build/bin/examples/Draw` and `build/bin/Resources` but got the same error:

```
> cd ~/Desktop/o3d_rd_test 
> ls
Draw		resources
> ./Draw 
libc++abi: terminating due to uncaught exception of type std::runtime_error: [Open3D Error] (std::string (anonymous namespace)::FindResourcePath(int, const char **)) /Users/cwr/Documents/code/Open3D/cpp/open3d/visualization/gui/Application.cpp:96: Could not find resource directory.

zsh: abort      ./Draw
```

My half-formed idea was that perhaps `FindResourcePath()` could be changed so that when it get to the end, before     `open3d::utility::LogError("Could not find resource directory.");` it could supply the path to `build/bin/Resources` as a fallback in the case where the user would prefer using the default resources over a hard error.

**Comment by ssheorey** [comment]
We do distribute resources, along with the Open3D binaries (see the release page on Github). `FindResourcePath` is called by Application::Initialize, which is called indirectly by O3DVisualizer. If you are building from source, you can print out all the directories that `FindResourcePath` is looking for on your platform to debug this. [Actually I think we should add that in the error message anyway]

I think the right way to do this is always have the resources directory in a few fixed locations wrt to the executable or Open3D DLL. Here's a SO question showing that this can be somewhat complex. BUt I think one of the solutions from there should work for us (C++17, no dependence on Boost, Qt, etc.).

https://stackoverflow.com/questions/1528298/get-path-of-executable

Also, we have to account for cases of system install, where executables (/usr/local/bin), DSOs (/usr/local/lib) and resources (/usr/local/share) are in different system folders.

**Reply by cwreynolds** [reply]
@ssheorey — everything you say seems reasonable to me, including about “the right way to do this.” But I’m unsure I can contribute anything useful to this issue in the near future. 

I did built my Open3D library from sources, build and run on a MacBook Pro. I have not been modifying Open3D. (My project is a layer on top of it.) I will try poking around, starting with your suggestion to trace where `FindResourcePath()` is looking, and perhaps a more detailed error message.

Still, it seems worth noting that `test2()` above requires external files to run. While for the same scene `test()` does not. It would be nice if these simplistic graphics could be drawn with inline procedural resources.
